testsuite/tests/typing-unique/unique_analysis.ml

```diff
@@ -0,0 +1,465 @@
+
+|}]
+
+(* for some reason the following error message is missing "another use". Don't
+know what's wrong *)
+let mark_top_shared =
+  let unique_ xs = 2 :: 3 :: [] in
+  match xs with
+  | x :: xx ->
+      let _ = unique_id xs in
+      unique_ xx
+  | [] -> []
+[%%expect{|
+Line 5, characters 24-26:
+5 |       let _ = unique_id xs in
+                            ^^
+Error: This is used uniquely here so cannot be used twice.  Another use is
+  The latter is the position-1 element of the constructor "::" in the value which is the former
+|}]
+
+let mark_top_shared =
+  let unique_ xs = 2 :: 3 :: [] in
+  let _ = unique_id xs in
+  match xs with
+  | x :: xx -> unique_ xx
+  | [] -> []
+[%%expect{|
+Line 3, characters 20-22:
+3 |   let _ = unique_id xs in
+                        ^^
+Error: This is used uniquely here so cannot be used twice.  Another use is
+  The latter is the position-1 element of the constructor "::" in the value which is the former
+|}]
+
+let mark_shared_in_one_branch b x =
+  if b then unique_id (x, 3.0)
+       else (x, x)
+[%%expect{|
+val mark_shared_in_one_branch : bool -> unique_ float -> float * float =
+  <fun>
+|}]
+
+let mark_shared_in_one_branch b x =
+  if b then (x, x)
+       else unique_id (x, 3.0)
+[%%expect{|
+val mark_shared_in_one_branch : bool -> unique_ float -> float * float =
+  <fun>
+|}]
+
+let expr_tuple_match f x y =
+  match f x, y with
+  | (a, b), c -> unique_ (a, c)
+[%%expect{|
+val expr_tuple_match : ('a -> unique_ 'b * 'c) -> 'a -> unique_ 'd -> 'b * 'd =
+  <fun>
+|}]
+
+let expr_tuple_match f x y =
+  match f x, y with
+  | (a, b) as t, c -> let d = unique_id t in unique_ (c, d)
+[%%expect{|
+val expr_tuple_match :
+  ('a -> unique_ 'b * 'c) -> 'a -> unique_ 'd -> 'd * ('b * 'c) = <fun>
+|}]
+
+let expr_tuple_match f x y =
+  match f x, y with
+  | (a, b) as t, c -> let d = unique_id t in unique_ (a, d)
+[%%expect{|
+Line 3, characters 40-41:
+3 |   | (a, b) as t, c -> let d = unique_id t in unique_ (a, d)
+                                            ^
+Error: This is used uniquely here so cannot be used twice.  Another use is
+Line 3, characters 54-55:
+3 |   | (a, b) as t, c -> let d = unique_id t in unique_ (a, d)
+                                                          ^
+  The latter is the position-0 element of the tuple which is the former
+|}]
+
+let tuple_parent_marked a b =
+  match (a, b) with
+  | (_, b) as _t -> shared_id b
+[%%expect{|
+val tuple_parent_marked : 'a -> 'b -> 'b = <fun>
+|}]
+
+let tuple_parent_marked a b =
+  match (a, b) with
+  | (true, b') -> unique_id b'
+  | (false, b') as _t -> shared_id b'
+[%%expect{|
+Line 2, characters 12-13:
+2 |   match (a, b) with
+                ^
+Error: This is used uniquely here so cannot be used twice. It is in a tuple matched against a variable. Another use is
+Line 3, characters 28-30:
+3 |   | (true, b') -> unique_id b'
+                                ^^
+
+|}]
```

Comment:
This example is an error because of `as _t` causes the usage of both `a` and `b`, which clashes with the unique use of `b` in the first branch. While intuitively this still shouldn't be an error, because the construction of `_t` (and hence the usage of `a` and `b`) happens only in the second branch which shouldn't interfere with the first branch, the current implementation views this differently. Upon seeing this `as _t`, both `a` and `b` will be used,  a new tree root is generated corresponding to `_t`, and further patterns `(false, b')` will be matched against this new tree root. It is therefore difficult to separate the usage caused by `false` and the usage caused by `_t`.
A better approach might be somehow to link `_t` to the paths of `a` and `b`. 

For now I will add a comment to ask for future improvement.